### ISSUE TITLE
Visualizing Top Tweeps with t-SNE in Javascript

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2338]
Visualizing Top Tweeps with t-SNE, in Javascript<br>
<br>
http://bit.ly/1EGGD6p<br>
<br>
<br>
<br>

<div>
<article>
  <p><img src="http://karpathy.github.io/assets/tsne_preview.jpeg"></p>

<p>I was recently looking into various ways of embedding unlabeled, high-dimensional data in 2 dimensions for visualization. A wide variety of methods have been proposed for this task. <a href="http://homepage.tudelft.nl/19j49/Matlab_Toolbox_for_Dimensionality_Reduction_files/TR_Dimensiereductie.pdf">This Review paper</a> from 2009 contains nice references to many of them (PCA, Kernel PCA, Isomap, LLE, Autoencoders, etc.). If you have Matlab available, the <a href="http://homepage.tudelft.nl/19j49/Matlab_Toolbox_for_Dimensionality_Reduction.html">Dimensionality Reduction Toolbox</a> has a nice implementation of many of these methods. Scikit Learn also has a brief section on <a href="http://scikit-learn.org/stable/modules/manifold.html">Manifold Learning</a> along with the implementation.</p>

<p>Among these algorithms, t-SNE comes across as one that has a pleasing, intuitive formulation, simple gradient and nice properties. Here is a <a href="http://www.youtube.com/watch?v=RJVL80Gg3lA">Google Tech Talks video</a> of Laurens van der Maaten (the author) explaining the method. I set out to re-implement t-SNE from scratch since doing so is the best way of learning something that I know of, and what better language to do this in than - Javascript! :)</p>

<p>Long story short, I’ve implemented t-SNE in JS, released it as <a href="https://github.com/karpathy/tsnejs">tsnejs on Github</a>, and created a small demo that uses the library to visualize the top twitter accounts based on what they talk about. In this post, I thought it might be fun to document a small 1-day project like this, from beginning to end. This also gives me an opportunity to describe some of my projects toolkit, which others might find useful.</p>

<h3 id="final-demo">Final demo</h3>
<p>First, take a look at the <a href="http://cs.stanford.edu/people/karpathy/tsnejs/">final demo</a>. To create this demo I found the top 500 most followed accounts on Twitter, downloaded 200 of their tweets and then measured differences in what they tweet about. These differences are then fed to t-SNE to produce a 2-dimensional visualization, where nearby people tweet similar things. Fun!</p>

<h3 id="fetching-top-tweeps">Fetching top tweeps</h3>
<p>We first have to identify the top 500 tweeps. I googled “top twitter accounts” and found http://twitaholic.com/ , which lists them out. However, the accounts are embedded in the webpage and we need to extract them in structured format. For this, I love a recent YC startup <a href="https://www.kimonolabs.com/">Kimono</a>; I use it extensively to scrape structured data from websites. It lets you click the elements of interest (the Twitter handles in this case), and extracts them out in JSON. Easy as pie!</p>

<h3 id="collecting-tweets">Collecting tweets</h3>
<p>Now we have a list of top 500 tweeps and we’d like to obtain their tweets to get an idea about what they tweet about. My library of choice for this task is <a href="https://github.com/tweepy/tweepy">Tweepy</a>. Their documentation is quite terrible but if you browse the source code things seem relatively simple. Here’s an example call to get 200 tweets for a given user:</p>

<div><div><pre><code><span>tweets</span> <span>=</span> <span>tweepy</span><span>.</span><span>Cursor</span><span>(</span><span>api</span><span>.</span><span>user_timeline</span><span>,</span> <span>screen_name</span><span>=</span><span>user</span><span>)</span><span>.</span><span>items</span><span>(</span><span>200</span><span>)</span>
</code></pre></div></div>

<p>We iterate this over all users, extract the tweet text, and dumpt it all into files, one per account. I had to be careful with two annoyances in process:</p>

<ul>
  <li>Twitter puts severe rate limits on API calls, so this actually took several hours to collect, wrapped up in try catch blocks and <code>time.sleep</code> calls.</li>
  <li>The returned text is in Unicode, which leads to trouble if you’re going to try to write it to file.</li>
</ul>

<p>One solution for the second annoyance is to use the codecs library:</p>

<div><div><pre><code><span>import</span> <span>codecs</span>
<span>codecs</span><span>.</span><span>open</span><span>(</span><span>filename</span><span>,</span> <span>'w'</span><span>,</span> <span>'utf-8'</span><span>)</span><span>.</span><span>write</span><span>(</span><span>tweet_texts</span><span>)</span>
</code></pre></div></div>

<p>Oh, and lets also grab and save the Twitter profile pictures, which we’ll use in the visualization. An example for one user might be:</p>

<div><div><pre><code><span>import</span> <span>urllib</span> <span># yes I know this is deprecated</span>
<span>userobj</span> <span>=</span> <span>api</span><span>.</span><span>get_user</span><span>(</span><span>screen_name</span> <span>=</span> <span>user</span><span>)</span>
<span>urllib</span><span>.</span><span>urlretrieve</span><span>(</span><span>imgname</span><span>,</span> <span>userobj</span><span>.</span><span>profile_image_url</span><span>)</span> <span># save image to disk</span>
</code></pre></div></div>

<p>I should mention that I write a lot of quick and dirty Python code in <a href="http://ipython.org/notebook.html">IPython Notebooks</a>, which I very warmly recommend. If you’re writing all your Python in text editors, you’re seriously missing out.</p>

<h3 id="quantifying-tweep-differences">Quantifying Tweep differences</h3>
<p>We now have 500 tweeps and their 200 most recent tweets concatenated in 500 files. We’d now like to find who tweets about similar things. <a href="http://scikit-learn.org/stable/">Scikit learn</a> is very nice for quick NLP tasks like this. In particular, we load up all the files and create a 500-long array where every element are the 200 concatenated tweets. Then we use the <a href="http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html">TfidfVectorizer</a> class to extract all words and bigrams from the text data, and to turn every user’s language into one tfidf vector. This vector is a fingerprint of the language that each person uses. Here’s how we can simply wire this up:</p>

<div><div><pre><code><span>from</span> <span>sklearn.feature_extraction.text</span> <span>import</span> <span>CountVectorizer</span><span>,</span><span>TfidfVectorizer</span>
<span>vectorizer</span> <span>=</span> <span>TfidfVectorizer</span><span>(</span><span>min_df</span><span>=</span><span>2</span><span>,</span> <span>stop_words</span> <span>=</span> <span>'english'</span><span>,</span>\
<span>strip_accents</span> <span>=</span> <span>'unicode'</span><span>,</span> <span>lowercase</span><span>=</span><span>True</span><span>,</span> <span>ngram_range</span><span>=</span><span>(</span><span>1</span><span>,</span><span>2</span><span>),</span>\
<span>norm</span><span>=</span><span>'l2'</span><span>,</span> <span>smooth_idf</span><span>=</span><span>True</span><span>,</span> <span>sublinear_tf</span><span>=</span><span>False</span><span>,</span> <span>use_idf</span><span>=</span><span>True</span><span>)</span>
<span>X</span> <span>=</span> <span>vectorizer</span><span>.</span><span>fit_transform</span><span>(</span><span>user_language_array</span><span>)</span>
<span>D</span> <span>=</span> <span>-</span><span>(</span><span>X</span> <span>*</span> <span>X</span><span>.</span><span>T</span><span>)</span><span>.</span><span>todense</span><span>()</span> <span># Distance matrix: dot product between tfidf vectors</span>
</code></pre></div></div>

<p>In the above, <code>user_language_array</code> is the 500-element array that has the concatenated tweets. The <code>TfidfVectorizer</code> class looks through all tweets and takes note of all words (unigrams) and word bigrams (i.e. series of two words). It builds a dictionary out of all unigram/bigrams and essentially counts up how often every person uses each one. Here’s an example of some tweet text converted to unigram/bigrams:</p>

<p><img src="http://karpathy.github.io/assets/tsne_sentprepro.jpeg"></p>

<p>The tfidf vectors are returned stacked up as rows inside <code>X</code>, which has size <code>500 x 87,342</code>. Every one of the 87,342 dimensions corresponds to some unigram or bigram. For example, the 10,000th dimension could correspond to the frequency of usage of the unigram “YOLO”. The vectors are L2 normalized, so the dot product between these vectors is related to the angle between any two vectors. This can be interpreted as the similarity of language. Finally, we dump the matrix and the usernames into a JSON file, and we’re ready to load things up in Javascript!</p>

<h3 id="the-visualization-parts">The Visualization parts</h3>
<p>We now create an .html file and import <a href="http://jquery.com/">jQuery</a> (as always), and <a href="http://d3js.org/">d3js</a>, which I like to use for any kind of plotting. We load up the JSON that stores our distances and usernames with jQuery, and use d3js to initialize the SVG element that will hold all the users. For starters, we plot the users at random position but we will soon arrange them so that similar users cluster nearby with t-SNE. Inspect the code on the <a href="http://cs.stanford.edu/people/karpathy/tsnejs/">demo page</a> to see the jQuery and d3js parts (Ctrl+U). In the code, we see a few things I like to use:</p>

<ul>
  <li>I like to use <strong>Google Fonts</strong> to get prettier-than-default fonts. Here, for example I’m importing Roboto, and then using it in the CSS.</li>
  <li>Next, we see an import of <strong>syntaxhighlighter</strong> code which dynamically highlights code on your page.</li>
  <li>Then we see <strong>Google tracking JS code</strong>, which lets me track statistics for the website on Google Analytics.</li>
  <li>I didn’t use <strong>Bootstrap</strong> on this website because it’s very small and simple, but normally I would because this makes your website right away work nicely on mobile.</li>
</ul>

<h3 id="t-sne">t-SNE</h3>

<p><img src="http://karpathy.github.io/assets/tsne_eg.jpeg"></p>

<p>Finally we get to the meat! We need to arrange the users in our d3js plot so that similar users appear nearby. The t-SNE cost function was described in this <a href="http://jmlr.csail.mit.edu/papers/volume9/vandermaaten08a/vandermaaten08a.pdf">2008 paper by van der Maaten and Hinton</a>. Similar to many other methods, we set up two distance metrics in the original and the embedded space and minimize their difference. In t-SNE in particular, the original space distance is based on a Gaussian distribution and the embedded space is based on the heavy-tailed Student-t distribution. The KL-divergence formulation has the nice property that it is asymmetric in how it penalizes distances between the two spaces:</p>

<ul>
  <li>If two points are close in the original space, there is a strong attractive force between the the points in the embedding</li>
  <li>Conversely, if the two points are far apart in the original space, the algorithm is relatively free to place these points around.</li>
</ul>

<p>Thus, the algorithm preferentially cares about preserving the local structure of the high-dimensional data. Conveniently, the authors link to multiple implementations of t-SNE on <a href="http://homepage.tudelft.nl/19j49/t-SNE.html">their website</a>, which allows us to see some code for reference as well (if you’re like me, reading code can be much easier than reading text descriptions). We’re ready to write up the Javascript version!</p>

<p>The final code can be seen in <a href="https://github.com/karpathy/tsnejs">tsne.js file</a>, on Github. Note how we’re wrapping all the JS code into a function closure so that we don’t pollute the global namespace. This is a very common trick in Javascript that is essentially used to implement classes. Note also the large number of utility boring code I had to include up top because Javascript is not exactly intended for math :) The core function where all magic happens is <code>costGrad()</code>, which computes the cost function and the gradient of the objective. The correct implementation of this function is double checked with <code>debugGrad()</code> gradient check. Once the analytic gradient checks out compared to numeric gradient, we’re good to go! We set up a piece of Javascript to call our <code>step()</code> function repeatedly (<code>setInterval()</code> call), and we plot the solution as it gets computed.</p>

<p>Phew! Final result, again: <a href="http://cs.stanford.edu/people/karpathy/tsnejs/">t-SNE demo</a>.</p>

<p>I hope some of the references were useful. If you use tsnejs to embed some of your data, let me know!</p>

<h2 id="bonus-word-embedding-t-sne-visualization">Bonus: Word Embedding t-SNE Visualization</h2>

<p>I created another demo, this time to visualize word vector embeddings. Head <a href="http://cs.stanford.edu/people/karpathy/tsnejs/wordvecs.html">over here</a> to see it. The word embeddings are trained as described in this <a href="http://www.socher.org/index.php/Main/ImprovingWordRepresentationsViaGlobalContextAndMultipleWordPrototypes">ACL 2012 paper</a>.</p>

<p>The (unsupervised) objective function makes it so that words that are interchangable (i.e. occur in very similar surrounding context) are close in the embedding. This comes across in the visualization!</p>

  </article>

  
  
  
  
 
 </div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via karpathy.github.io http://bit.ly/1oRUNGz<br>
<br>
January 7, 2019 at 03:05PM
